User model and user routes with authentication

mattegan111 opened the pull request.

mattegan111 committed:
Add body parsing middleware

Allows us to parse json data sent in request body.

mattegan111 committed:
Add validation to 'Register User' route

User model is imported but not used.

mattegan111 committed:
Add User model

mattegan111 committed:
Encrypt User password using bcryptjs

mattegan111 committed:
Return JWT on register

Implement jsonwebtoken so a newly registered user can access protected routes.

mattegan111 committed:
Add auth login route

mattegan111 committed:
Add login route and auth middleware

nathanrobb commented on server/middleware/auth.js:
```diff
@@ -0,0 +1,24 @@
+
+		req.user = decoded.user;
+		next();
+	} catch (err) {
```
What happens if `next()` causes an error to be thrown? Is this current behaviour what we want?

Perhaps the `try catch` should be around the token `verify`?

mattegan111 replied:
For the sake of code organisation it might be nice to leave it inside the try, even if, upon it causing an error, the error message will be inaccurate. It's a parameterless function that can actually be used to handle errors so I feel we're unlikely to have problems with this line.

Keen to put lines 5-11 inside the try.

nathanrobb replied:
Have a look at this doc, [https://expressjs.com/en/guide/error-handling.html](https://expressjs.com/en/guide/error-handling.html).
We might not need any general try catches, as express would do that for us by default (potentially).

mattegan111 replied:
In this case I still believe a try/catch is most appropriate however I did find a way to reduce the use of try/catch in the routes, see link below.
https://thecodebarbarian.com/80-20-guide-to-express-error-handling

I would like to cover operational errors for this project but am not concerned with handling programming errors as it's just a bit out of scope.

nathanrobb commented on server/routes/users.js:
```diff
@@ -1,11 +1,88 @@
+				password,
+			});
+
+			const salt = await bcrypt.genSalt(10);
+
+			user.password = await bcrypt.hash(password, salt);
```
1. Whats the benefit of using a `salt` when hashing a password?
2. Why are we  _hashing_ here (instead of encrypting)?
3. How does the authenticate endpoint verify the password? The salt is only referenced here?

No changes, this current code is good, just asking questions.

mattegan111 replied:
Pasting our Discord conversation. :)

1. Prevents rainbow table attacks where hashes of common passwords are compared with hashed values stored in our database.
2. Because we won't be decrypting the hashed value, we just use the stored hashed value as-is for comparison only.
3. The value bcrypt gives us to save in our database contains a hash of both the salt and the password but also contains a hash of just the salt. So upon login it takes the user input password and hashes it, takes the hashed salt from the db. Hashes those together. Compares that value with the part of the value in the db which represents the hash of both salt and password.

nathanrobb commented on server/routes/users.js:
```diff
@@ -1,11 +1,88 @@
+				payload,
+				config.get('jwtSecret'),
+				{
+					expiresIn: 360000, //TODO change to 3600 for production
```
Do you have an environments file yet? This would be a good addition to that?
Or in the config where the `jwtsecret` is stored?

mattegan111 replied:
Sounds good, will implement one of these.

mattegan111 replied:
Note: Added to config.

nathanrobb commented on server/routes/auth.js:
```diff
@@ -1,18 +1,91 @@
 const express = require('express');
 const router = express.Router();
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
```
Are these new packages from NPM? If so, are you missing a commit to your `package.json`.

EDIT: Oh, they were already added back in 442009b ✅

nathanrobb commented on server/routes/auth.js:
```diff
@@ -1,18 +1,91 @@
+		const user = await User.findById(req.user.id).select(
+			'-password'
+		);
+		res.json(user);
```
Prob not worth, but did we need to verify the user id exists in the db? What if they were deleted?

mattegan111 replied:
If there's no user, I think `select` will fail and the catch will run. Will have to check, otherwise yeah may as well check if exists.

mattegan111 replied:
I was wrong, `select` doesn't trigger the catch. Added an if statement to check for and handle a null user value.

nathanrobb commented on server/models/User.js:
```diff
@@ -0,0 +1,48 @@
+	bio: {
+		type: String,
+	},
+	friends: {
+		type: Array,
+	},
+	notifications: {
+		type: Array,
+	},
+	receivedfriendrequests: {
+		type: Array,
+	},
+	sentfriendrequests: {
+		type: Array,
+	},
+	post: {
+		type: Array,
+	},
+	comments: {
+		type: Array,
+	},
```
No change needed. But perhaps these would be better to commit when adding each of these other tables.

It makes the commits easier to read as all the table to table links are included in the same commit/PR.

Right now these are a little strange to read as they do nothing (yet).

nathanrobb commented on server/routes/users.js:
```diff
@@ -1,11 +1,88 @@
+				.json({ errors: errors.array() });
+		}
+
+		const { firstname, lastname, email, password } =
+			req.body;
```
No way to set a bio. Is this API intended to just to the bare minimum to make a user and later there will be a webpage to edit/set details on the user? If so then no change needed.

nathanrobb reviewed (approved):
Nice, looks good. Mostly just questions.

mattegan111 committed:
Fix missing async

Function uses await so must be made async.

mattegan111 committed:
Moved README

mattegan111 committed:
Add error catching middleware

Reduce use of try/catch to improve code organisation.

mattegan111 committed:
Check user exists when getting logged in user

Incase user is deleted or not found in database for whatever reason, we want to return an error.

nathanrobb commented on server/middleware/errorCatcher.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = function errorCatcher(fn) {
+	return (req, res, next) => {
+		fn(req, res, next).catch(next);
```
so passing the caught error into `next` triggers the error handling middleware defined in `server.js`?

Nice, I think this makes it a bit nicer to keep each route's code fairly specific to the operation at hand with the small tradeoff of _magic_ via the `errorCatcher` function.

I like this approach personally.

nathanrobb commented on server/routes/auth.js:
```diff
@@ -21,6 +21,11 @@ router.get(
 		const user = await User.findById(req.user.id).select(
 			'-password'
 		);
+		if (!user) {
```
Does the `findById` return `null` or the `select`?

If it's the find then we could get a null reference error here.

mattegan111 replied:
The `findById` returns null if no user is found, however the `select` doesn't cause an error when this happens. (Not sure why, but `findById` works with async features under the hood, could be something to do with this???)

I found a better solution anyway which removes the need for select entirely, by setting the User schema not to select the password field by default.

https://stackoverflow.com/questions/12096262/how-to-protect-the-password-field-in-mongoose-mongodb-so-it-wont-return-in-a-qu

I also noticed `const user` should be `let user` because we define it throughout the file. It only worked here as it's the first instance of `user`, then elsewhere we're using `'let user`.

nathanrobb replied:
> I also noticed const user should be let user because we define it throughout the file.

I think we want to keep `const` here. It's only being set once in each method scope (`const` is scoped to this route handle function), the other routes would define their own user scoped to their own route handler.

I think the other `user` object below should also be changed to `const` as it's only wanting to be set once within its scope.

`let` would be used if we were setting the `user` object multiple times within the same method scope. If we were using `var` we would be having a slightly different convo, but in this case `let` and `const` are per route handler scoped and some the fact they are defined in other handlers is irrelevant.

mattegan111 committed:
Use const appropriately

mattegan111 committed:
Prevent password select by default

Prevent password selection at a schema level so this behaviour only needs to be defined once in code and is therefore less likely to be missed in future.
